src/basic/io-util.c

```diff
@@ -125,21 +128,39 @@ int loop_write(int fd, const void *buf, size_t nbytes, bool do_poll) {
                 p = buf;
         }
 
+        end = timestamp_is_set(timeout) ? usec_add(now(CLOCK_MONOTONIC), timeout) : USEC_INFINITY;
```

Comment:
hmm so, this would mean specifiying timeout=0 would mean infinity timeout. I think it would make more sense to to make it mean "write, but never wait". i.e. that we try to write something, and if we are about to sleep we just fail.